convex/aiTown/agent.ts

```diff
@@ -0,0 +1,371 @@
+          const messageUuid = crypto.randomUUID();
+          conversation.setIsTyping(now, player, messageUuid);
+          this.startOperation(game, now, 'agentGenerateMessage', {
+            worldId: game.worldId,
+            playerId: player.id,
+            agentId: this.id,
+            conversationId: conversation.id,
+            otherPlayerId: otherPlayer.id,
+            messageUuid,
+            type: 'leave',
+          });
+          return;
+        }
+        // Wait for the awkward deadline if we sent the last message.
+        if (conversation.lastMessage.author === player.id) {
+          const awkwardDeadline = conversation.lastMessage.timestamp + AWKWARD_CONVERSATION_TIMEOUT;
+          if (now < awkwardDeadline) {
+            return;
+          }
+        }
+        // Wait for a cooldown after the last message to simulate "reading" the message.
+        const messageCooldown = conversation.lastMessage.timestamp + MESSAGE_COOLDOWN;
+        if (now < messageCooldown) {
+          return;
+        }
+        // Grab the lock and send a message!
+        console.log(`${player.id} continuing conversation with ${otherPlayer.id}.`);
+        const messageUuid = crypto.randomUUID();
+        conversation.setIsTyping(now, player, messageUuid);
+        this.startOperation(game, now, 'agentGenerateMessage', {
+          worldId: game.worldId,
+          playerId: player.id,
+          agentId: this.id,
+          conversationId: conversation.id,
+          otherPlayerId: otherPlayer.id,
+          messageUuid,
+          type: 'continue',
+        });
+        return;
+      }
+    }
+  }
+
+  startOperation<Name extends keyof AgentOperations>(
+    game: Game,
+    now: number,
+    name: Name,
+    args: Omit<FunctionArgs<AgentOperations[Name]>, 'operationId'>,
+  ) {
+    if (this.inProgressOperation) {
+      throw new Error(
+        `Agent ${this.id} already has an operation: ${JSON.stringify(this.inProgressOperation)}`,
+      );
+    }
+    const operationId = crypto.randomUUID();
+    console.log(`Agent ${this.id} starting operation ${name} (${operationId})`);
+    game.scheduleOperation(name, { operationId, ...args } as any);
+    this.inProgressOperation = {
+      name,
+      operationId,
+      started: now,
+    };
+  }
+
+  serialize(): SerializedAgent {
+    return {
+      id: this.id,
+      playerId: this.playerId,
+      toRemember: this.toRemember,
+      lastConversation: this.lastConversation,
+      lastInviteAttempt: this.lastInviteAttempt,
+      inProgressOperation: this.inProgressOperation,
+    };
+  }
+}
+
+export const serializedAgent = {
+  id: agentId,
+  playerId: playerId,
+  toRemember: v.optional(conversationId),
+  lastConversation: v.optional(v.number()),
+  lastInviteAttempt: v.optional(v.number()),
+  inProgressOperation: v.optional(
+    v.object({
+      name: v.string(),
+      operationId: v.string(),
+      started: v.number(),
+    }),
+  ),
```

Comment:
It's non-ideal that I have to define the same state in 4 places. since the validator is required, others can flow from that, with the mapping to `GameId` in there as the outlier? Maybe Agent has a single variable "state" that is of type SerializedAgent (where we get SerializedAgent to have id as GameId). we can follow up on this. Ideally if you added a type to one of these places the other ones would complain but that's only true for these last two